Raise error rather than overwrite data when sharding is not respected

robertwb opened the pull request:
This is particularly relevant when withoutSharding() is used due to
[BEAM-159].

peihe commented:
R: @peihe

Do we want to fail in constructName()?
Could you add a unit test with the fileNamingTemplate that does not disambiguate numFiles?

dhalperi commented:
This seems like it's doing two different things.

I think the fix you want could be applied [here](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextIO.java#L641): just throw if numShards is set (or however we detect this). And similarly in AvroIO -- these are the only places that have a notion of sharding in Beam.

I think the check you've added here is great but we should make the error message more appropriate.

dhalperi commented:
R: @dhalperi

robertwb commented:
We can't fail there as we need to support this for the Dataflow and Direct runners. 

Do you have a better suggestion for an error message?

dhalperi commented:
That code should not be reached for Dataflow and Direct runners.

robertwb commented:
I thought we didn't intercept apply() calls during pipeline construction
depending on the runner anymore.

On Wed, Jun 8, 2016 at 4:40 PM, Daniel Halperin <notifications@github.com>
wrote:

> That code should not be reached for Dataflow and Direct runners.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-beam/pull/436#issuecomment-224763627>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAdqgYLtkqcDZK3lebFCQ-Ihb5fL6es4ks5qJ1MAgaJpZM4IxczJ>
> .
>

dhalperi commented:
My suggested error message:

```java
int numDistinctShards = new HashSet<String>(destFilenames).size();
checkState(numDistinctShards == numFiles,
    "Shard name template '%s' only generated %d distinct file names for %d files.",
    fileNamingTemplate, numDistinctShards, numFiles);
```

robertwb commented:
The difficulty with this error message is that it's not very actionable.
The root cause is (almost certainly) lack of support for numShards.

On Wed, Jun 8, 2016 at 4:45 PM, Daniel Halperin <notifications@github.com>
wrote:

> My suggested error message:
>
> int numDistinctShards = new HashSet<String>(destFilenames).size();
> checkState(numDistinctShards == numFiles,
>     "Shard name template '%s' only generated %d distinct file names for %d files.",
>     fileNamingTemplate, numDistinctShards, numFiles);
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-beam/pull/436#issuecomment-224764413>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAdqge90pWowzMHo5GrQcZCaMC0hcgSpks5qJ1QjgaJpZM4IxczJ>
> .
>

dhalperi commented:
No way -- I think you may have caught a real bug here, if the user provided a shard template like `SSS-of-NNN` and we generated 10K files. That's worth calling out on its own.

robertwb commented:
The number of digits is a lower, not upper, bound so 10K files would be fine. 

I've changed the error, the most important thing to me is that it's not a silent data loss. Also added a test.

dhalperi commented:
LGTM, thanks.
